--- ../sonobe_mod/folding-schemes/src/folding/foldpot/utils.rs
+++ ../sonobe_mod/folding-schemes/src/folding/foldpot/utils.rs
@@ -736,6 +736,7 @@
 pub fn write_g16_optimized_bn254(path: &Path, pk: &ark_groth16::ProvingKey<Bn254>, vk: &ark_groth16::VerifyingKey<Bn254>) {
     let b_debug = true;
+    let start = std::time::Instant::now();
     let path_str = path.to_str().unwrap();
 
     let meta_path = format!("{}.meta", path_str);
     let file = File::create(&meta_path).expect("create meta err");
     let mut writer = BufWriter::new(file);
@@ -747,6 +748,8 @@
     pk.vk.alpha_g1.serialize_compressed(&mut writer).expect("ser pk.vk.alpha_g1");
     pk.beta_g1.serialize_compressed(&mut writer).expect("ser pk.beta_g1");
     pk.delta_g1.serialize_compressed(&mut writer).expect("ser pk.delta_g1");
+    
+    drop(writer); // drop writer to flush the file before potential read
 
     serialize_affines_raw(&vk.gamma_abc_g1, &format!("{}_vk_gamma_abc_g1", path_str));
     serialize_affines_raw(&pk.a_query, &format!("{}_pk_a_query", path_str));
@@ -770,7 +773,7 @@
         total_size += metadata(&format!("{}_vy.data", prefix)).map(|m| m.len()).unwrap_or(0);
         total_size += metadata(&format!("{}_vb.data", prefix)).map(|m| m.len()).unwrap_or(0);
     }
-    println!("PERF 1003: [write_g16_optimized_bn254] elements: {}, size: {} bytes", pk.a_query.len(), total_size);
+    println!("PERF 1003: [write_g16_optimized_bn254] path: {:?}, elements: {}, size: {} bytes, time: {:?}", path, pk.a_query.len(), total_size, start.elapsed());
 
     if b_debug {
         let (pk_read, vk_read) = read_g16_optimized_bn254(path);
@@ -782,6 +785,7 @@
 
 pub fn read_g16_optimized_bn254(path: &Path) -> (ark_groth16::ProvingKey<Bn254>, ark_groth16::VerifyingKey<Bn254>) {
+    let start = std::time::Instant::now();
     let path_str = path.to_str().unwrap();
 
     let meta_path = format!("{}.meta", path_str);
@@ -826,7 +830,7 @@
         total_size += metadata(&format!("{}_vy.data", prefix)).map(|m| m.len()).unwrap_or(0);
         total_size += metadata(&format!("{}_vb.data", prefix)).map(|m| m.len()).unwrap_or(0);
     }
-    println!("PERF 1003: [read_g16_optimized_bn254] elements: {}, size: {} bytes", pk.a_query.len(), total_size);
+    println!("PERF 1003: [read_g16_optimized_bn254] path: {:?}, elements: {}, size: {} bytes, time: {:?}", path, pk.a_query.len(), total_size, start.elapsed());
 
     (pk, vk)
 }
