--- ../sonobe_mod/folding-schemes/src/folding/foldpot/utils.rs
+++ ../sonobe_mod/folding-schemes/src/folding/foldpot/utils.rs
@@ -675,13 +675,13 @@
     
     // Convert to uncompressed coordinates
-    v.par_iter().enumerate().for_each(|(i, pt)| {
+    vx.par_iter_mut().zip(vy.par_iter_mut()).zip(vb.par_iter_mut()).zip(v.par_iter()).for_each(|(((vx_i, vy_i), vb_i), pt)| {
         if pt.is_zero() {
-            vx[i] = zero;
-            vy[i] = zero;
-            vb[i] = true;
+            *vx_i = zero;
+            *vy_i = zero;
+            *vb_i = true;
         } else {
             // Unwrapping is safe because the point is not zero
-            vx[i] = *pt.x().unwrap();
-            vy[i] = *pt.y().unwrap();
-            vb[i] = false;
+            *vx_i = *pt.x().unwrap();
+            *vy_i = *pt.y().unwrap();
+            *vb_i = false;
         }
